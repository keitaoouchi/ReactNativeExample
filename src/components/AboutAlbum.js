import React from "react";
import {FlatList, View} from "react-native";
import {Actions} from "react-native-router-flux";
import AlbumHeader from "./AlbumHeader";
import TrackCell from "./TrackCell";
import Workaround from "../misc/Workaround";
import Color from "../misc/Color";
import AboutAlbumActionCreator from "../flux/action/AboutAlbumActionCreator";
import AboutAlbumStore from "../flux/store/AboutAlbumStore";
import {Dispatcher} from "flux";
import TrackPlayer from "./../misc/TrackPlayer";

const AboutAlbumPresentation = ({tracks, album, albumDetail, artist, playingTrack, actionCreator}) => (
  <View style={ style.root }>
    <FlatList
      enableEmptySections={true}
      data={tracks}
      keyExtractor={(item, index) => index}
      // FIXME: ListHeaderComponent would re-mount every time FlatList re-render it's item.
      // https://github.com/facebook/react-native/issues/13365
      ListHeaderComponent={() =>
        <AlbumHeader
          key={1}
          album={album}
          albumDetail={albumDetail}
          onPressArtistName={() => {
            Actions.artist({
              title: artist.name,
              album: album,
              artist: artist,
              hideNavBar: false
            });
          }}
        />
      }
      renderItem={({item}) => (
        <TrackCell
          track={item}
          onPress={() => actionCreator.toggleTrack(item)}
          isPlaying={playingTrack === item}
        />
      )}
    />
  </View>
);

export default class AboutAlbum extends React.Component {

  constructor(props) {
    super(props);

    const dispatcher = new Dispatcher();
    this.actionCreator = new AboutAlbumActionCreator(dispatcher);
    this.store = new AboutAlbumStore(dispatcher);
    this.player = TrackPlayer.shared;
  }

  componentDidMount() {
    this.store.addListener(() => this.forceUpdate());
    this.playerListener = this.player.store.addListener(() => this.forceUpdate());
    this.actionCreator.fetchAlbumDetail(this.props.album);
  }

  componentWillUnmount() {
    this.playerListener.remove();
  }

  render() {
    const state = this.store.getState();
    const playerState = this.player.store.getState();
    return <AboutAlbumPresentation
      tracks={state.albumDetail ? state.albumDetail.tracks : []}
      album={this.props.album}
      albumDetail={state.albumDetail}
      artist={this.props.album.artist}
      playingTrack={playerState.playingTrack}
      actionCreator={this.player.actionCreator}
    />;
  }
}

const style = {
  root: {flex: 1, backgroundColor: Color.black, ...Workaround.style.root},
};
