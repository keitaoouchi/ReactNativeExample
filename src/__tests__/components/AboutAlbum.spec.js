import "react-native";
import React from "react";
import renderer from "react-test-renderer";
import AboutAlbum from "../../../src/components/AboutAlbum";
import AlbumFactory from "../factories/AlbumFactory";
import FetchMock from "fetch-mock";

describe('AboutAlbum', () => {

  beforeAll(() => {
    this.album = AlbumFactory.build('AlbumFactory');
    FetchMock.get('*', {});
  });

  it('renders correctly', () => {
    const tree = renderer.create(
      <AboutAlbum album={this.album}/>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});