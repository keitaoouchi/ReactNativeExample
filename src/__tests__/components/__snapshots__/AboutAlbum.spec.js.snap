// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AboutAlbum renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#181818",
      "flex": 1,
      "marginTop": 64,
    }
  }
>
  <RCTScrollView
    ListHeaderComponent={[Function]}
    data={Array []}
    disableVirtualization={false}
    enableEmptySections={true}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onViewableItemsChanged={undefined}
    removeClippedSubviews={true}
    renderItem={[Function]}
    renderScrollComponent={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            }
          }
        >
          <Image
            onLoadEnd={[Function]}
            source={
              Object {
                "uri": "https://example.com/1",
              }
            }
            style={
              Object {
                "height": 240,
                "margin": 20,
                "opacity": 0,
                "width": 240,
              }
            }
          />
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "white",
                "fontSize": 24,
                "fontWeight": "bold",
                "margin": 12,
              }
            }
          >
            AlbumTitle-1
          </Text>
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "center",
                "margin": 8,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "white",
                  "fontSize": 16,
                  "opacity": 0.5,
                }
              }
            >
              By
            </Text>
            <View
              accessibilityComponentType={undefined}
              accessibilityLabel={undefined}
              accessibilityTraits={undefined}
              accessible={true}
              hitSlop={undefined}
              isTVSelectable={true}
              onLayout={undefined}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                  "padding": 8,
                }
              }
              testID={undefined}
              tvParallaxProperties={undefined}
            >
              <View>
                <Text
                  accessible={true}
                  allowFontScaling={true}
                  ellipsizeMode="tail"
                  style={
                    Object {
                      "color": "white",
                      "fontSize": 16,
                      "fontWeight": "normal",
                    }
                  }
                >
                  Name-1
                </Text>
              </View>
            </View>
          </View>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "white",
                "fontSize": 12,
                "margin": 8,
                "opacity": 0.5,
              }
            }
          >
            
          </Text>
          <View
            accessibilityComponentType={undefined}
            accessibilityLabel={undefined}
            accessibilityTraits={undefined}
            accessible={true}
            hitSlop={undefined}
            isTVSelectable={true}
            onLayout={undefined}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID={undefined}
            tvParallaxProperties={undefined}
          >
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "#1ed760",
                    "borderRadius": 22,
                    "height": 44,
                    "margin": 12,
                    "padding": 8,
                    "width": 200,
                  },
                  false,
                ]
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "color": "white",
                    "fontSize": 16,
                    "fontWeight": "bold",
                    "margin": 4,
                    "textAlign": "center",
                  }
                }
              >
                Spotifyで聴く
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
